weave-js/src/components/PagePanelComponents/PublishModal.tsx

```diff
@@ -0,0 +1,219 @@
+import {
+  Dropdown,
+  DropdownItemProps,
+  Form,
+  Input,
+  Modal,
+} from 'semantic-ui-react';
+import React from 'react';
+import styled from 'styled-components';
+import {RED_550} from '@wandb/weave/common/css/color.styles';
+import {WBButton} from '@wandb/weave/common/components/elements/WBButtonNew';
+import * as query from './Home/query';
+import {useIsAuthenticated} from './util';
+import {useEffect, useState} from 'react';
+import * as w from '@wandb/weave/core';
+import {useNodeValue} from '@wandb/weave/react';
+import {Icon} from '../Icon';
+import {TakeActionType} from './persistenceStateMachine';
+import * as M from './Modal.styles';
+
+const Error = styled.div`
+  font-size: 14px;
+  color: ${RED_550};
+`;
+Error.displayName = 'S.Error';
+
+type PublishModalProps = {
+  defaultName: string | null;
+  open: boolean;
+  acting: boolean;
+  takeAction: TakeActionType;
+  onClose: () => void;
+};
+
+// TODO: Starting/ending with '.' or '-' should probably not be allowed.
+const isValidBoardName = (name: string) => {
+  return (
+    0 < name.length && name.length <= 128 && /^[a-zA-Z0-9_.-]+$/.test(name)
+  );
+};
+
+export const PublishModal = ({
+  defaultName,
+  open,
+  acting,
+  takeAction,
+  onClose,
+}: PublishModalProps) => {
+  const [boardName, setBoardName] = useState(defaultName ?? '');
+  const [entityName, setEntityName] = useState('');
+  const [projectName, setProjectName] = useState('');
+  const isValidName = isValidBoardName(boardName);
+  const showError = boardName.length > 0 && !isValidName;
+
+  const isAuthenticated = useIsAuthenticated();
+  const userEntities = query.useUserEntities(isAuthenticated);
+  const userName = query.useUserName(isAuthenticated);
+
+  const iconStyle = {
+    verticalAlign: 'middle',
+    display: 'inline-block',
+    marginRight: 8,
+  };
+
+  const entityOptions: DropdownItemProps[] =
+    userEntities.result.length === 0
+      ? []
+      : userEntities.result.sort().map(entName => ({
+          icon: (
+            <Icon
+              name={
+                entName === userName.result
+                  ? 'user-profile-personal'
+                  : 'users-team'
+              }
+              style={iconStyle}
+            />
+          ),
+          value: entName,
+          text: entName,
+        }));
+
+  const projectsNode = w.opEntityProjects({
+    entity: w.opRootEntity({
+      entityName: w.constString(entityName),
+    }),
+  });
+  const projectDataNode = w.opMap({
```

Comment:
This weave expression can be simplified to `w.opProjectName({project: projectsNode})`